cni/network/network.go

```diff
@@ -485,51 +452,59 @@ func (plugin *NetPlugin) Add(args *cniSkel.CmdArgs) error {
 	if plugin.ipamInvoker == nil {
 		switch nwCfg.Ipam.Type {
 		case network.AzureCNS:
-			cnsURL := "http://localhost:" + strconv.Itoa(cnsPort)
-			cnsClient, er := cnscli.New(cnsURL, defaultRequestTimeout)
+			cnsClient, er := cnscli.New("", defaultRequestTimeout)
 			if er != nil {
 				return fmt.Errorf("initializing cns client failed with err %w", er)
 			}
-			plugin.ipamInvoker = NewCNSInvoker(k8sPodName, k8sNamespace, cnsClient)
+			plugin.ipamInvoker = NewCNSInvoker(k8sPodName, k8sNamespace, cnsClient, nil)
 
 		default:
-			plugin.ipamInvoker = NewAzureIpamInvoker(plugin, &nwInfo)
-		}
-	}
+			if nwCfg.MultiTenancy {
+				cnsClient, er := cnscli.New(nwCfg.CNSUrl, defaultRequestTimeout)
+				if er != nil {
+					return fmt.Errorf("failed to create cns client for multitenancy %w", er)
+				}
 
-	// Allocate from azure ipam
-	if !nwCfg.MultiTenancy {
-		result, resultV6, err = plugin.ipamInvoker.Add(nwCfg, args, &subnetPrefix, options)
-		if err != nil {
-			return err
+				plugin.report.Context = "AzureCNIMultitenancy"
+				plugin.multitenancyClient.Init(cnsClient, AzureNetIOShim{})
+				plugin.ipamInvoker = NewCNSInvoker(k8sPodName, k8sNamespace, cnsClient, plugin.multitenancyClient)
+			} else {
+				plugin.ipamInvoker = NewAzureIpamInvoker(plugin, &nwInfo)
+			}
 		}
+	}
 
-		defer func() {
-			if err != nil {
-				plugin.cleanupAllocationOnError(result, resultV6, nwCfg, args, options)
-			}
-		}()
+	ipamAddResult, err = plugin.ipamInvoker.Add(IPAMAddOpt{nwCfg: nwCfg, args: args, options: options})
+	if err != nil {
+		return fmt.Errorf("IPAM Invoker Add failed with error: %w", err)
 	}
 
+	defer func() {
+		if err != nil && !nwCfg.MultiTenancy {
+			plugin.cleanupAllocationOnError(ipamAddResult.ipv4Result, ipamAddResult.ipv6Result, nwCfg, args, options)
+		}
+	}()
+
 	// Create network
 	if nwInfoErr != nil {
 		// Network does not exist.
 		log.Printf("[cni-net] Creating network %v.", networkID)
-		if nwInfo, err = plugin.createNetworkInternal(networkID, policies, args, nwCfg, cnsNetworkConfig, subnetPrefix, result, resultV6); err != nil {
-			log.Errorf("Create network failed:%w", err)
+		if nwInfo, err = plugin.createNetworkInternal(networkID, policies, args, nwCfg, ipamAddResult.ncResponse,
+			*ipamAddResult.subnetPrefix, ipamAddResult.ipv4Result, ipamAddResult.ipv6Result); err != nil {
+			log.Errorf("Create network failed: %w", err)
 			return err
 		}
 
-		log.Printf("[cni-net] Created network %v with subnet %v.", networkID, subnetPrefix.String())
+		log.Printf("[cni-net] Created network %v with subnet %v.", networkID, ipamAddResult.subnetPrefix.String())
 	}
 
-	natInfo := getNATInfo(nwCfg.ExecutionMode, options[network.SNATIPKey], nwCfg.MultiTenancy, enableSnatForDns)
+	natInfo := getNATInfo(nwCfg.ExecutionMode, options[network.SNATIPKey], nwCfg.MultiTenancy, ipamAddResult.enableSnatForDNS)
```

Comment:
can we just pas nwCfg as arg instead  of nwCfg.ExecutionMode and nwCfg.MultiTenancy?

Reply:
This was already agreed on the last PR, not sure if we need to pass the whole nwCfg here